internal/client.go

```diff
@@ -6,13 +6,98 @@ import (
 	"fmt"
 	"io"
 	"mime"
+	"net"
 	"net/http"
 	"net/url"
 	"path"
 	"strings"
 	"unicode"
 )
 
+// Discover performs a DNS-based CalDAV/CardDAV service discovery as described in
+// RFC 6764 section 6. It returns the URL to the CalDAV/CardDAV server.
+func Discover(service string, host string) (string, error) {
+	if service != "caldav" && service != "carddav" {
+		return "", fmt.Errorf("webdav: service discovery of type %v not supported", service)
+	}
+
+	path := ""
+
+	// Check for SRV records for the service we want, only lookup secure versions
+	// (caldavs, carddavs), plaintext connections are insecure
+	_, addrs, err := net.LookupSRV(fmt.Sprintf("%vs", service), "tcp", host)
+	if dnsErr, ok := err.(*net.DNSError); ok {
+		if dnsErr.IsTemporary {
+			return "", err
+		}
+	} else if err != nil {
+		return "", err
+	}
+
+	if len(addrs) > 0 {
+		srvTarget := strings.TrimSuffix(addrs[0].Target, ".")
+
+		// If we found one, check for TXT records specifying the path
+		if srvTarget != "" {
+			txtRecs, err := net.LookupTXT(fmt.Sprintf("_%vs._tcp.%v", service, host))
+			if dnsErr, ok := err.(*net.DNSError); ok {
+				if dnsErr.IsTemporary {
+					return "", err
+				}
+			} else if err != nil {
+				return "", err
+			}
+
+			for _, txtRec := range txtRecs {
+				// This is not correct according to RFC 6763, but LookupTXT merges all constituent strings together
+				for _, txtRecKeyVal := range strings.Split(txtRec, " ") {
+					if strings.HasPrefix(txtRecKeyVal, "path=") {
+						path = strings.TrimPrefix(txtRecKeyVal, "path=")
+						break
+					}
+				}
+
+				if path != "" {
+					break
+				}
+			}
+
+			if addrs[0].Port == 443 {
+				host = srvTarget
+			} else {
+				host = fmt.Sprintf("%v:%v", srvTarget, addrs[0].Port)
+			}
+		}
+	}
+
+	// If we didn't get a path from TXT records, use the default well-known location
+	if path == "" {
+		path = fmt.Sprintf("/.well-known/%v", service)
+	}
+
+	u := url.URL{Scheme: "https", Host: host, Path: path}
+	serviceUrl := u.String()
+
+	// Check if the resulting URL hosts a service
+	req, err := http.NewRequest(http.MethodOptions, serviceUrl, nil)
```

Comment:
[RFC 6764 section 6](https://tools.ietf.org/html/rfc6764#section-6) mandates that a PROPFIND request should be made with the username/password set.

Alps has a different requirement: it needs to sanity check the URLs on startup (without any username/password).

To follow the RFC, we could:

- Take an `webdav.HTTPClient` to handle auth
- Change OPTIONS to a PROPFIND request with a `DAV:current-user-principal` prop
- Return the `DAV:current-user-principal` prop

Reply:
To accomodate for Alps, we could have an intermediate `DiscoverContextURL` function that follows the RFC up to and not including the PROPFIND request.